### ISSUE TITLE
Do post-processing on the output

### ISSUE THREAD

**Issue by amirfz** [Aggregate-Intellect/practical-llms#28]
there are a few areas:

- if code is generated ensure that it runs
- if urls are provided ensure that they open

**Comment by babasahebpinjar** [comment]
- if code is generated ensure that it runs

This is a tricky one, we can explore to use services like **replit** to send code via api and see if it executes. But how many programming languages are we going to support for this ? 

- if urls are provided ensure that they open: 

We can use requests library to check if the url generated is valid.

def is_valid_url(url):
    try:
        response = requests.head(url)
        # Check if the response status code indicates success (2xx) or redirection (3xx)
        if response.status_code // 100 in [2, 3]:
            return True
        else:
            return False
    except requests.RequestException:
        return False

url_to_check = "https://www.example.com"
if is_valid_url(url_to_check):
    print(f"The URL '{url_to_check}' is valid.")
else:
    print(f"The URL '{url_to_check}' is not valid or cannot be reached.")

**Comment by amirfz** [comment]
what to do if URL is invalid:

- remove the URL, OR
- find a way to fix it (eg. google search + set limits on acceptable edit distance ratio)

**Comment by babasahebpinjar** [comment]
for the second point

- find a way to fix it (eg. google search + set limits on acceptable edit distance ratio)

I don't think we should try to fix it. The url is generated, the OpenAI API itself does not disclose the details of the sources it uses to generate responses to user queries.

If we fix it then by giving a different url then the answer may not be same in the new url

@amirfz when you say  "if the URL is provided in the context" do we have access to this context ?

**Comment by amirfz** [comment]
@babasahebpinjar but don't we know the URL that is used to fetch the information using the Google search or our Retrieval tool? it should be safe to use that URL. I agree that we shouldn't use a random URL that we are not certain if it is related

**Comment by amirfz** [comment]
modify the prompt to use the URLs provided by the tools in post-processing

**Comment by oshoma** [comment]
A few thoughts here:

1. Checking a URL to be sure it's valid, a la `is_valid_url(url)` above, is actually tricky for many reasons, including:
a) URL redirects to another URL (or multiple URLs) and the ultimate URL at end of chain is not valid 
b) a server is temporarily down or slow to respond, so the GET request fails
c) a service like Cloudflare blocks the client for some reason
In situations like this you may suspect the URL is invalid, but can't be sure without retrying later or following redirects. Retries must be done asynchronously. 

2. A related issue is safety. You might learn that a URL is "bad" for some reason, e.g. contains malware. To protect against this I would use a service like Google's [Web Risk](https://cloud.google.com/web-risk/docs) or [Safe Browsing API](https://developers.google.com/safe-browsing/v4/lookup-api) (non-commercial use only). I think this is worth implementing sometime as a separate feature, provided the Tools don't already do it themselves.

To summarize, there are more actions one can take to be more certain of a URL's validity, but some of them are too slow to be done synchronously while the user waits for a response. 

One idea is to display a visual warning next to URLs that haven't been validated.
